### ISSUE TITLE
Target specific client on clustered socket.io server and using acknowledgment callback

### ISSUE THREAD

**Issue by peteruithoven** [socketio/socket.io#1811]
I've clustered my socket.io server and I'm looking for a way to emit to a specific client and receiving a acknowledgment callback. The problem that the server doesn't have a reference to a instance of the target client, because this client can be connected to another socket.io server instance (because it's clustered). 
I know I can emit to a room named after the the socket.id or a custom room where I put the client in. But the problem is that this is considered broadcasting and then I can't use a acknowledgment callback. 
I don't want to add complexity to the target client. The target client should just be able to call the callback. 

I see two possible solutions: 
- Find a way to create a fake socket instance with the socket.id (so I can use existing callback code)
- Broadcast to a room and build my own callback system, see: https://github.com/Automattic/socket.io/issues/1656#issuecomment-49077695

A fake socket instance wouldn't actually have a connection to a client, but I hope to use it's emit function, have it talk to it's adapter (socket.io-redis) and have it receive the acknowledgment callback. 

Any tips on these solutions, maybe something I should also consider.

**Comment by sgiachero** [comment]
Has this scenario been solved using socket.io, socket.io-redis, and socket.io-emitter?

**Comment by darrachequesne** [comment]
This was eventually implemented in version `4.5.0` (https://github.com/socketio/socket.io/commit/8b204570a94979bbec307f23ca078f30f5cf07b0):

```js
io.timeout(1000).emit("some-event", (err, responses) => {
  // ...
});
```

Feedback is welcome!

**Comment by rituraj12797** [comment]
@FREEZX   @peteruithoven  @sgiachero  @darrachequesne   @maccman 

recently faced the same issue i think . like i am operating on a clustered environment using NodeJS clusters , what initially i was doing is that 
  a. i will transmit the message with the data of the room of recipient and the socketID of recipient 
       but the problem was that when i was sending this message to the room via the IPC provided by cluster adapter 
       the message is broadcasted to al the clients on reaching the room due to the code 
       
       `socket.to(`${room_no}1).emit("message","<some_message>)` 
       
   b.  eventually i figured out that somehow just by writing this code 
   ` socket.to(`${socketId}`).emit("message", {text});`
   
   it would automatically search for the client with the given socketID globally ( idk how is this working if anyone could help ) 
   and automatically transmits the message from the sender (in room a) to the receiver (in room b) eliminating the need of finding the room.
   
   still i would like to know if anyone could help me finding how is it able to globally search for the recipient how efficient is it
